File: synth_ai/sdk/task/trace_helpers.py
"""Helpers for trace correlation IDs and trace payloads in task apps.

This module provides utilities for task apps to:
1. Validate trace_correlation_id presence
2. Include trace_correlation_id in rollout responses (top-level, metadata, trace)
3. Build v3 trace payloads for trace-only responses

NOTE: As of the contracts update, trace_correlation_id is now a REQUIRED field
in RolloutRequest. For new code, simply use request.trace_correlation_id directly.

See monorepo/trace_creation_and_judgement.txt "Fatal Guards" section for requirements.
"""

import logging
from datetime import UTC
from typing import Any

logger = logging.getLogger(__name__)


def validate_trace_correlation_id(
    trace_correlation_id: str | None,
    policy_config: dict[str, Any] | None = None,
    fatal: bool = False,
    *,
    run_id: str | None = None,  # Deprecated, for logging only
) -> str | None:
    """
    Validate that trace_correlation_id is present.

    NOTE: With the updated contracts, trace_correlation_id is now a REQUIRED
    field in RolloutRequest. This function is primarily for validating
    extraction from legacy sources (inference URLs).

    Args:
        trace_correlation_id: The correlation ID to validate
        policy_config: Policy configuration for debugging (optional)
        fatal: If True, raise ValueError on missing ID. If False, log error only.
        run_id: DEPRECATED - Only used for logging. Use trace_correlation_id for identification.

    Returns:
        trace_correlation_id if present, None if missing (when fatal=False)

    Raises:
        ValueError: If trace_correlation_id is missing and fatal=True
    """
    if not trace_correlation_id:
        inference_url = (
            policy_config.get("inference_url", "NOT_SET") if policy_config else "NOT_SET"
        )
        id_for_log = run_id or trace_correlation_id or "UNKNOWN"
        error_msg = (
            "CRITICAL: Cannot extract trace_correlation_id!\n"
            "\n"
            f"ID: {id_for_log}\n"
            f"Inference URL: {inference_url}\n"
            "\n"
            "Checked:\n"
            "1. inference_url path segments\n"
            "2. inference_url query params\n"
            "\n"
            "Task app CANNOT proceed without trace_correlation_id.\n"
            "This indicates the RL trainer is not sending it correctly.\n"
            "\n"
            "See monorepo/trace_creation_and_judgement.txt 'Fatal Guards' section.\n"
        )

        if fatal:
            raise ValueError(error_msg)
        else:
            logger.error(error_msg)

    return trace_correlation_id


def include_trace_correlation_id_in_response(
    response_data: dict[str, Any],
    trace_correlation_id: str | None,
    run_id: str | None = None,
) -> dict[str, Any]:
    """
    Include trace_correlation_id in all required locations of rollout response.

    Required locations:
    1. Top-level response["trace_correlation_id"]
    2. response["trace"]["metadata"]["trace_correlation_id"] (and session_trace metadata if present)

    Args:
        response_data: RolloutResponse dict (from .model_dump())
        trace_correlation_id: The correlation ID to include
        run_id: DEPRECATED - Only used for logging. Use trace_correlation_id for identification.

    Returns:
        Modified response_data with trace_correlation_id in all required places
    """
    id_for_log = run_id or trace_correlation_id or "UNKNOWN"
    if not trace_correlation_id:
        logger.error(
            "include_trace_correlation_id_in_response: missing trace_correlation_id "
            "for id=%s - cannot include in response",
            id_for_log,
        )
        return response_data

    # 1. Add to top-level (REQUIRED)
    if "trace_correlation_id" not in response_data:
        response_data["trace_correlation_id"] = trace_correlation_id
        logger.debug(
            "include_trace_correlation_id: added to top-level id=%s cid=%s",
            id_for_log,
            trace_correlation_id,
        )

    # 2. Add to trace metadata (REQUIRED)
    trace_block = response_data.get("trace")
    if isinstance(trace_block, dict):
        trace_meta = trace_block.get("metadata")
        if not isinstance(trace_meta, dict):
            trace_meta = {}
            trace_block["metadata"] = trace_meta
        if "trace_correlation_id" not in trace_meta:
            trace_meta["trace_correlation_id"] = trace_correlation_id
        corr_ids = trace_meta.get("correlation_ids")
        corr_map = dict(corr_ids) if isinstance(corr_ids, dict) else {}
        corr_map.setdefault("trace_correlation_id", trace_correlation_id)
        trace_meta["correlation_ids"] = corr_map

        session_trace = trace_block.get("session_trace")
        if isinstance(session_trace, dict):
            session_meta = session_trace.get("metadata")
            if not isinstance(session_meta, dict):
                session_meta = {}
                session_trace["metadata"] = session_meta
            session_meta.setdefault("trace_correlation_id", trace_correlation_id)

    logger.debug(
        "include_trace_correlation_id: completed id=%s cid=%s "
        "added to top-level and trace metadata",
        id_for_log,
        trace_correlation_id,
    )

    return response_data


def build_trace_payload(
    messages: list[dict[str, Any]],
    response: dict[str, Any] | None = None,
    *,
    correlation_id: str | None = None,
    session_id: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """
    Build a v3 trace payload with event_history for trace-only responses.

    Args:
        messages: The messages sent to the LLM (input)
        response: The LLM response dict (output). Should include 'choices' or 'content'.
        correlation_id: Trace correlation ID (from ?cid= param)
        session_id: Optional session ID for the trace
        metadata: Optional additional metadata

    Returns:
        A trace dict with event_history suitable for RolloutResponse.trace
    """
    import uuid
    from datetime import datetime

    event_history: list[dict[str, Any]] = []

    llm_response: dict[str, Any] = {}
    if isinstance(response, dict):
        if "message" in response:
            llm_response = dict(response)
        elif (
            "choices" in response
            and isinstance(response.get("choices"), list)
            and response["choices"]
        ):
            first_choice = (
                response["choices"][0] if isinstance(response["choices"][0], dict) else {}
            )
            llm_response = {
                "message": first_choice.get("message") if isinstance(first_choice, dict) else {},
                "usage": response.get("usage", {}),
                "finish_reason": first_choice.get("finish_reason")
                if isinstance(first_choice, dict)
                else None,
            }
        else:
            llm_response = dict(response)

    llm_event: dict[str, Any] = {
        "type": "lm_call",
        "event_type": "lm_call",
        "timestamp": datetime.now(UTC).isoformat(),
        "llm_request": {"messages": messages},
        "llm_response": llm_response,
    }

    # Add correlation ID if provided
    if correlation_id:
        llm_event["correlation_id"] = correlation_id

    event_history.append(llm_event)

    trace_metadata: dict[str, Any] = dict(metadata or {})
    trace_metadata.setdefault("session_id", session_id or str(uuid.uuid4()))
    if correlation_id:
        trace_metadata.setdefault("trace_correlation_id", correlation_id)
        corr_ids = trace_metadata.get("correlation_ids")
        corr_map = dict(corr_ids) if isinstance(corr_ids, dict) else {}
        corr_map.setdefault("trace_correlation_id", correlation_id)
        trace_metadata["correlation_ids"] = corr_map

    trace: dict[str, Any] = {
        "schema_version": "3.0",
        "event_history": event_history,
        "markov_blanket_message_history": [],
        "metadata": trace_metadata,
    }

    logger.debug(
        "build_trace_payload: created trace with %d events, session_id=%s, cid=%s",
        len(event_history),
        trace_metadata.get("session_id"),
        correlation_id,
    )

    return trace


def build_trajectory_trace(
    messages: list[dict[str, Any]],
    response: dict[str, Any] | None = None,
    *,
    correlation_id: str | None = None,
    session_id: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Backward-compatible alias for build_trace_payload."""

    return build_trace_payload(
        messages=messages,
        response=response,
        correlation_id=correlation_id,
        session_id=session_id,
        metadata=metadata,
    )


def include_event_history_in_response(
    response_data: dict[str, Any],
    messages: list[dict[str, Any]] | None = None,
    response: dict[str, Any] | None = None,
    *,
    run_id: str,
    correlation_id: str | None = None,
) -> dict[str, Any]:
    """
    Ensure response.trace includes a v3 event_history payload.

    Args:
        response_data: RolloutResponse dict (from .model_dump())
        messages: Messages for the LLM call (for building event_history)
        response: LLM response payload
        run_id: Rollout run_id for logging
        correlation_id: Trace correlation ID

    Returns:
        Modified response_data with event_history in response.trace
    """
    trace_block = response_data.get("trace")
    if not isinstance(trace_block, dict):
        trace_block = {}
        response_data["trace"] = trace_block

    event_history = trace_block.get("event_history")
    session_trace = trace_block.get("session_trace")
    if not event_history and isinstance(session_trace, dict):
        event_history = session_trace.get("event_history")

    if isinstance(event_history, list) and event_history:
        return response_data

    new_trace = build_trace_payload(
        messages=messages or [],
        response=response,
        correlation_id=correlation_id,
        metadata={"run_id": run_id},
    )

    # Merge new trace payload into the existing trace block.
    trace_meta = trace_block.get("metadata")
    if isinstance(trace_meta, dict):
        merged_meta = dict(new_trace.get("metadata", {}))
        merged_meta.update(trace_meta)
        trace_block["metadata"] = merged_meta
    else:
        trace_block["metadata"] = new_trace.get("metadata", {})

    trace_block.setdefault("schema_version", new_trace.get("schema_version"))
    trace_block["event_history"] = new_trace.get("event_history", [])
    trace_block.setdefault(
        "markov_blanket_message_history",
        new_trace.get("markov_blanket_message_history", []),
    )

    if isinstance(session_trace, dict) and "event_history" not in session_trace:
        session_trace["event_history"] = trace_block["event_history"]

    logger.debug(
        "include_event_history_in_response: added event_history run_id=%s events=%d",
        run_id,
        len(trace_block.get("event_history", [])),
    )
    return response_data


def include_event_history_in_trajectories(
    response_data: dict[str, Any],
    messages_by_trajectory: list[list[dict[str, Any]]] | None = None,
    responses_by_trajectory: list[dict[str, Any]] | None = None,
    *,
    run_id: str,
    correlation_id: str | None = None,
) -> dict[str, Any]:
    """Backward-compatible alias for include_event_history_in_response."""

    messages = messages_by_trajectory[0] if messages_by_trajectory else None
    response = responses_by_trajectory[0] if responses_by_trajectory else None
    return include_event_history_in_response(
        response_data,
        messages=messages,
        response=response,
        run_id=run_id,
        correlation_id=correlation_id,
    )


def verify_trace_correlation_id_in_response(
    response_data: dict[str, Any],
    expected_correlation_id: str | None,
    run_id: str | None = None,
) -> bool:
    """
    Verify that trace_correlation_id is present in all required locations.

    Args:
        response_data: RolloutResponse dict to verify
        expected_correlation_id: The correlation ID that should be present
        run_id: DEPRECATED - Only used for logging. Use trace_correlation_id for identification.

    Returns:
        True if all required locations have the correlation ID, False otherwise
    """
    id_for_log = run_id or expected_correlation_id or "UNKNOWN"
    if not expected_correlation_id:
        logger.error(
            "verify_trace_correlation_id: no expected_correlation_id provided for id=%s", id_for_log
        )
        return False

    errors = []

    # Check top-level
    if response_data.get("trace_correlation_id") != expected_correlation_id:
        errors.append(
            f"Top-level missing or mismatch: "
            f"expected={expected_correlation_id} actual={response_data.get('trace_correlation_id')}"
        )

    # Check trace metadata
    trace_block = response_data.get("trace")
    trace_meta_id = None
    if isinstance(trace_block, dict):
        trace_meta = trace_block.get("metadata")
        if isinstance(trace_meta, dict):
            trace_meta_id = trace_meta.get("trace_correlation_id")
        if trace_meta_id != expected_correlation_id:
            session_trace = trace_block.get("session_trace")
            if isinstance(session_trace, dict):
                session_meta = session_trace.get("metadata")
                if isinstance(session_meta, dict):
                    trace_meta_id = session_meta.get("trace_correlation_id")
        if trace_meta_id != expected_correlation_id:
            errors.append(
                "trace.metadata missing or mismatch: "
                f"expected={expected_correlation_id} actual={trace_meta_id}"
            )

    if errors:
        logger.error("verify_trace_correlation_id: FAILED id=%s\n%s", id_for_log, "\n".join(errors))
        return False

    logger.debug(
        "verify_trace_correlation_id: PASSED id=%s cid=%s", id_for_log, expected_correlation_id
    )
    return True

File: synth_ai_core/src/localapi/trace_helpers.rs
use crate::errors::CoreError;
use chrono::Utc;
use serde_json::{Map, Value};
use url::Url;
use uuid::Uuid;

fn parse_url(inference_url: &str) -> Option<Url> {
    if let Ok(parsed) = Url::parse(inference_url) {
        return Some(parsed);
    }
    if inference_url.starts_with('/') {
        let candidate = format!("http://localhost{}", inference_url);
        return Url::parse(&candidate).ok();
    }
    if !inference_url.contains("://") {
        let candidate = format!("http://{}", inference_url);
        return Url::parse(&candidate).ok();
    }
    None
}

pub fn extract_trace_correlation_id(inference_url: Option<&str>) -> Option<String> {
    let raw = inference_url?.trim();
    if raw.is_empty() {
        return None;
    }

    let parsed = parse_url(raw)?;
    let path_segments: Vec<&str> = parsed.path().split('/').filter(|s| !s.is_empty()).collect();

    if path_segments.len() >= 3 {
        let tail = &path_segments[path_segments.len() - 2..];
        if tail == ["chat", "completions"] {
            if let Some(segment) = path_segments.get(path_segments.len() - 3) {
                if segment.starts_with("trace_")
                    || segment.starts_with("cid_")
                    || segment.starts_with("eval_")
                {
                    return Some(segment.to_string());
                }
            }
        }
    }

    for segment in path_segments.iter().rev() {
        if segment.starts_with("trace_")
            || segment.starts_with("cid_")
            || segment.starts_with("eval_")
        {
            return Some(segment.to_string());
        }
    }

    let query = parsed.query().unwrap_or("");
    for (key, value) in url::form_urlencoded::parse(query.as_bytes()) {
        let key = key.to_string();
        if key == "cid" || key == "trace_correlation_id" || key == "trace" {
            let val = value.to_string();
            if !val.trim().is_empty() {
                return Some(val);
            }
        }
    }

    None
}

pub fn validate_trace_correlation_id(
    trace_correlation_id: Option<&str>,
    inference_url: Option<&str>,
    fatal: bool,
) -> Result<Option<String>, CoreError> {
    if let Some(cid) = trace_correlation_id {
        if !cid.trim().is_empty() {
            return Ok(Some(cid.to_string()));
        }
    }

    let inference_url = inference_url.unwrap_or("NOT_SET");
    let error_msg = format!(
        "CRITICAL: Cannot extract trace_correlation_id!\n\nID: UNKNOWN\nInference URL: {}\n\nChecked:\n1. inference_url path segments\n2. inference_url query params\n\nTask app CANNOT proceed without trace_correlation_id.\nThis indicates the RL trainer is not sending it correctly.\n\nSee monorepo/trace_creation_and_judgement.txt 'Fatal Guards' section.\n",
        inference_url
    );

    if fatal {
        return Err(CoreError::InvalidInput(error_msg));
    }

    Ok(None)
}

pub fn include_trace_correlation_id_in_response(
    response_data: &Value,
    trace_correlation_id: Option<&str>,
) -> Value {
    let cid = match trace_correlation_id {
        Some(value) if !value.trim().is_empty() => value.trim(),
        _ => return response_data.clone(),
    };

    let mut output = response_data.clone();
    let obj = match output.as_object_mut() {
        Some(obj) => obj,
        None => return output,
    };

    obj.entry("trace_correlation_id".to_string())
        .or_insert_with(|| Value::String(cid.to_string()));

    if let Some(trace_val) = obj.get_mut("trace") {
        if let Some(trace_obj) = trace_val.as_object_mut() {
            let meta = trace_obj
                .entry("metadata".to_string())
                .or_insert_with(|| Value::Object(Map::new()));
            if let Some(meta_obj) = meta.as_object_mut() {
                meta_obj
                    .entry("trace_correlation_id".to_string())
                    .or_insert_with(|| Value::String(cid.to_string()));
                let corr_ids = meta_obj
                    .entry("correlation_ids".to_string())
                    .or_insert_with(|| Value::Object(Map::new()));
                if let Some(corr_map) = corr_ids.as_object_mut() {
                    corr_map
                        .entry("trace_correlation_id".to_string())
                        .or_insert_with(|| Value::String(cid.to_string()));
                }
            }

            if let Some(session_trace) = trace_obj.get_mut("session_trace") {
                if let Some(session_obj) = session_trace.as_object_mut() {
                    let session_meta = session_obj
                        .entry("metadata".to_string())
                        .or_insert_with(|| Value::Object(Map::new()));
                    if let Some(session_meta_obj) = session_meta.as_object_mut() {
                        session_meta_obj
                            .entry("trace_correlation_id".to_string())
                            .or_insert_with(|| Value::String(cid.to_string()));
                    }
                }
            }
        }
    }

    output
}

pub fn build_trace_payload(
    messages: &Value,
    response: Option<&Value>,
    correlation_id: Option<&str>,
    session_id: Option<&str>,
    metadata: Option<&Value>,
) -> Value {
    let messages_value = match messages {
        Value::Array(list) => Value::Array(list.clone()),
        _ => Value::Array(Vec::new()),
    };

    let llm_response = match response {
        Some(Value::Object(obj)) => {
            if obj.contains_key("message") {
                Value::Object(obj.clone())
            } else if let Some(Value::Array(choices)) = obj.get("choices") {
                if let Some(Value::Object(first)) = choices.get(0) {
                    let mut map = Map::new();
                    map.insert(
                        "message".to_string(),
                        first.get("message").cloned().unwrap_or(Value::Null),
                    );
                    map.insert(
                        "usage".to_string(),
                        obj.get("usage")
                            .cloned()
                            .unwrap_or(Value::Object(Map::new())),
                    );
                    map.insert(
                        "finish_reason".to_string(),
                        first.get("finish_reason").cloned().unwrap_or(Value::Null),
                    );
                    Value::Object(map)
                } else {
                    Value::Object(obj.clone())
                }
            } else {
                Value::Object(obj.clone())
            }
        }
        Some(other) => other.clone(),
        None => Value::Object(Map::new()),
    };

    let mut llm_event = Map::new();
    llm_event.insert("type".to_string(), Value::String("lm_call".to_string()));
    llm_event.insert(
        "event_type".to_string(),
        Value::String("lm_call".to_string()),
    );
    llm_event.insert(
        "timestamp".to_string(),
        Value::String(Utc::now().to_rfc3339()),
    );
    let mut llm_request = Map::new();
    llm_request.insert("messages".to_string(), messages_value);
    llm_event.insert("llm_request".to_string(), Value::Object(llm_request));
    llm_event.insert("llm_response".to_string(), llm_response);
    llm_event.insert("api_format".to_string(), Value::String("chat".to_string()));
    if let Some(cid) = correlation_id {
        if !cid.trim().is_empty() {
            llm_event.insert("correlation_id".to_string(), Value::String(cid.to_string()));
        }
    }

    let mut event_history = Vec::new();
    event_history.push(Value::Object(llm_event));

    let mut trace_meta = match metadata {
        Some(Value::Object(obj)) => obj.clone(),
        _ => Map::new(),
    };
    let session_id = session_id
        .map(|s| s.to_string())
        .unwrap_or_else(|| Uuid::new_v4().to_string());
    trace_meta
        .entry("session_id".to_string())
        .or_insert(Value::String(session_id));

    if let Some(cid) = correlation_id {
        if !cid.trim().is_empty() {
            trace_meta
                .entry("trace_correlation_id".to_string())
                .or_insert(Value::String(cid.to_string()));
            let corr_ids = trace_meta
                .entry("correlation_ids".to_string())
                .or_insert_with(|| Value::Object(Map::new()));
            if let Some(corr_map) = corr_ids.as_object_mut() {
                corr_map
                    .entry("trace_correlation_id".to_string())
                    .or_insert(Value::String(cid.to_string()));
            }
        }
    }

    let mut trace = Map::new();
    trace.insert(
        "schema_version".to_string(),
        Value::String("4.0".to_string()),
    );
    trace.insert("event_history".to_string(), Value::Array(event_history));
    trace.insert(
        "markov_blanket_message_history".to_string(),
        Value::Array(Vec::new()),
    );
    trace.insert("metadata".to_string(), Value::Object(trace_meta));

    Value::Object(trace)
}

pub fn build_trajectory_trace(
    messages: &Value,
    response: Option<&Value>,
    correlation_id: Option<&str>,
    session_id: Option<&str>,
    metadata: Option<&Value>,
) -> Value {
    build_trace_payload(messages, response, correlation_id, session_id, metadata)
}

pub fn include_event_history_in_response(
    response_data: &Value,
    messages: Option<&Value>,
    response: Option<&Value>,
    run_id: &str,
    correlation_id: Option<&str>,
) -> Value {
    let mut output = response_data.clone();
    let obj = match output.as_object_mut() {
        Some(obj) => obj,
        None => return output,
    };

    let trace_val = obj
        .entry("trace".to_string())
        .or_insert_with(|| Value::Object(Map::new()));
    let trace_obj = match trace_val.as_object_mut() {
        Some(obj) => obj,
        None => return output,
    };

    let mut event_history = trace_obj.get("event_history");
    if event_history.is_none() {
        if let Some(Value::Object(session_trace)) = trace_obj.get("session_trace") {
            event_history = session_trace.get("event_history");
        }
    }
    if matches!(event_history, Some(Value::Array(list)) if !list.is_empty()) {
        return output;
    }

    let meta_map = {
        let mut map = Map::new();
        map.insert("run_id".to_string(), Value::String(run_id.to_string()));
        Value::Object(map)
    };

    let new_trace = build_trace_payload(
        messages.unwrap_or(&Value::Array(Vec::new())),
        response,
        correlation_id,
        None,
        Some(&meta_map),
    );

    if let Some(new_trace_obj) = new_trace.as_object() {
        let incoming_meta = trace_obj.get("metadata");
        let mut merged_meta = match new_trace_obj.get("metadata") {
            Some(Value::Object(meta)) => meta.clone(),
            _ => Map::new(),
        };
        if let Some(Value::Object(existing)) = incoming_meta {
            for (k, v) in existing.iter() {
                merged_meta.insert(k.clone(), v.clone());
            }
        }
        trace_obj.insert("metadata".to_string(), Value::Object(merged_meta));
        if let Some(schema) = new_trace_obj.get("schema_version") {
            trace_obj
                .entry("schema_version".to_string())
                .or_insert_with(|| schema.clone());
        }
        if let Some(events) = new_trace_obj.get("event_history") {
            trace_obj.insert("event_history".to_string(), events.clone());
        }
        if let Some(blanket) = new_trace_obj.get("markov_blanket_message_history") {
            trace_obj
                .entry("markov_blanket_message_history".to_string())
                .or_insert_with(|| blanket.clone());
        }

        let events_clone = trace_obj.get("event_history").cloned();
        if let Some(Value::Object(session_trace)) = trace_obj.get_mut("session_trace") {
            if !session_trace.contains_key("event_history") {
                if let Some(events) = events_clone {
                    session_trace.insert("event_history".to_string(), events);
                }
            }
        }
    }

    output
}

pub fn include_event_history_in_trajectories(
    response_data: &Value,
    messages_by_trajectory: Option<&Value>,
    responses_by_trajectory: Option<&Value>,
    run_id: &str,
    correlation_id: Option<&str>,
) -> Value {
    let messages = messages_by_trajectory
        .and_then(|value| value.as_array())
        .and_then(|arr| arr.get(0));
    let response = responses_by_trajectory
        .and_then(|value| value.as_array())
        .and_then(|arr| arr.get(0));

    include_event_history_in_response(response_data, messages, response, run_id, correlation_id)
}

pub fn verify_trace_correlation_id_in_response(
    response_data: &Value,
    expected_correlation_id: Option<&str>,
) -> bool {
    let expected = match expected_correlation_id {
        Some(value) if !value.trim().is_empty() => value,
        _ => return false,
    };

    let obj = match response_data.as_object() {
        Some(obj) => obj,
        None => return false,
    };

    if obj.get("trace_correlation_id").and_then(|v| v.as_str()) != Some(expected) {
        return false;
    }

    let trace_val = match obj.get("trace") {
        Some(Value::Object(trace)) => trace,
        _ => return false,
    };

    let mut trace_meta_id = trace_val
        .get("metadata")
        .and_then(|v| v.as_object())
        .and_then(|meta| meta.get("trace_correlation_id"))
        .and_then(|v| v.as_str());

    if trace_meta_id != Some(expected) {
        if let Some(Value::Object(session_trace)) = trace_val.get("session_trace") {
            trace_meta_id = session_trace
                .get("metadata")
                .and_then(|v| v.as_object())
                .and_then(|meta| meta.get("trace_correlation_id"))
                .and_then(|v| v.as_str());
        }
    }

    trace_meta_id == Some(expected)
}
